presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java

```diff
@@ -793,6 +803,179 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
+                    s3.setThrowOnNextUpload();
+                }
+            });
+            assertDirEmpty(stagingDir);
+        }
+    }
+
+    @Test
+    public void testRoundTripWrite()
+            throws IOException, URISyntaxException
+    {
+        long minMultipartSize = 100L;
+        List<Integer> sizes = ImmutableList.of(99, 100, 101, 199, 200, 201);
+        for (Integer fileSize : sizes) {
+            byte[] data = new byte[fileSize];
+            ThreadLocalRandom.current().nextBytes(data);
+            try (PrestoS3FileSystem fs = new PrestoS3FileSystem()) {
+                MockAmazonS3 s3 = new MockAmazonS3();
+                Configuration conf = new Configuration();
+                conf.setLong(S3_MULTIPART_MIN_PART_SIZE, minMultipartSize);
+                java.nio.file.Path stagingDir = Files.createTempDirectory("presto-s3-abort-multipart-test");
+                conf.set(S3_STAGING_DIRECTORY, stagingDir.toString());
+                fs.initialize(new URI("s3n://test-bucket/"), conf);
+                fs.setS3Client(s3);
+                // verify for each write type the file is correct
+                // write(byte[])
+                assertRoundTripWrite(() -> {
+                    Path filePath = new Path("/test" + fileSize + "byteArray");
+                    try (FSDataOutputStream out = fs.create(filePath)) {
+                        out.write(data);
+                    }
+                    return filePath;
+                }, fs, data);
+                // write(byte)
+                assertRoundTripWrite(() -> {
+                    Path filePath = new Path("/test" + fileSize + "byte");
+                    try (FSDataOutputStream out = fs.create(filePath)) {
+                        for (byte bit : data) {
+                            out.write(bit);
+                        }
+                    }
+                    return filePath;
+                }, fs, data);
+                // write(byte[], index, len)
+                assertRoundTripWrite(() -> {
+                    Path filePath = new Path("/test" + fileSize + "byteChunk");
+                    try (FSDataOutputStream out = fs.create(filePath)) {
+                        int chunkSize = 20;
+                        for (int i = 0; i < fileSize / chunkSize; i++) {
+                            out.write(data, i * chunkSize, chunkSize);
+                        }
+                        out.write(data, (fileSize / chunkSize) * chunkSize, fileSize % chunkSize);
+                    }
+                    return filePath;
+                }, fs, data);
+            }
+        }
+    }
+
+    private static void assertRoundTripWrite(IOSupplier<Path> writerExpected, FileSystem fs, byte[] actual)
+            throws IOException
+    {
+        Path writtenFile = writerExpected.get();
+        byte[] input = new byte[actual.length];
+        try (FSDataInputStream in = fs.open(writtenFile)) {
+            assertEquals(actual.length, in.read(input));
```

Comment:
We're missing the case where the actually written data is larger than expected. How about we use this for the method -
```
private static void assertRoundTripWrite(IOSupplier<Path> writerExpected, FileSystem fs, byte[] expected)
            throws IOException
    {
        Path writtenFile = writerExpected.get();
        byte[] actual;
        try (FSDataInputStream in = fs.open(writtenFile)) {
            actual = ByteStreams.toByteArray(in);
        }
        assertEquals(actual, expected);
    }
```

Reply:
Thanks - good catch. I've updated the method